File: 167. Two Sum II - Input Array Is Sorted/167. Two Sum II - Input Array Is Sorted.js
function twoSum(numbers, target) {
    let left = 0;
    let right = numbers.length - 1;

    while (left < right) {
        const sum = numbers[left] + numbers[right];
        if (sum === target) {
            return [left + 1, right + 1]; // Return 1-based indices
        } else if (sum < target) {
            left++; // Move left pointer to the right
        } else {
            right--; // Move right pointer to the left
        }
    }

    // The problem guarantees exactly one solution, so no need for a return here
}

// Example usage:
console.log(twoSum([2, 7, 11, 15], 9)); // Output: [1, 2]
console.log(twoSum([2, 3, 4], 6)); // Output: [1, 3]
console.log(twoSum([-1, 0], -1)); // Output: [1, 2]

File: 167. Two Sum II - Input Array Is Sorted/167. Two Sum II - Input Array Is Sorted.rs
impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        let mut left = 0;
        let mut right = numbers.len() - 1;

        while left < right {
            let sum = numbers[left] + numbers[right];
            
            if sum == target {
                // We add 1 to convert from 0-indexed to 1-indexed
                return vec![(left + 1) as i32, (right + 1) as i32];
            } else if sum < target {
                left += 1;
            } else {
                right -= 1;
            }
        }

        // This line should never be reached given the problem constraints
        vec![]
    }
}

// Example usage and tests
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_1() {
        assert_eq!(Solution::two_sum(vec![2, 7, 11, 15], 9), vec![1, 2]);
    }

    #[test]
    fn test_example_2() {
        assert_eq!(Solution::two_sum(vec![2, 3, 4], 6), vec![1, 3]);
    }

    #[test]
    fn test_example_3() {
        assert_eq!(Solution::two_sum(vec![-1, 0], -1), vec![1, 2]);
    }
}